JSDK-2684 better heartbeat

makarandp0 opened the pull request:
update the signaling websocket connection heartbeat logic to:

- Each outgoing message is now treated as an outgoing heartbeat.
- Each incoming message is treated as an incoming heartbeat.
- The incoming heartbeat timeout is now 15 seconds, .
- Removed the offset added to the incoming heartbeat timeout.
- added offset of -200ms  200 to outgoing timeout to account for network latency.
- Add a timeout for getting the onopen event on the WebSocket (15 sec). After that, attempt to reconnect only if this is not the first attempt. 

- [x] TODO: verify that server is okay with not sending heartbeat.
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

manjeshbhargav commented on lib/twilioconnection.js:
```diff
@@ -30,14 +30,16 @@ const states = {
 const WS_CLOSE_HELLO_FAILED = 3002;
 const WS_CLOSE_SEND_FAILED = 3003;
 const WS_CLOSE_NETWORK_CHANGED = 3004;
+const WS_CLOSE_FAILED_TO_OPEN = 3005;
```
@makarandp0 Nitpick:
```suggestion
const WS_CLOSE_OPEN_TIMEOUT = 3005;
```

manjeshbhargav commented on lib/twilioconnection.js:
```diff
@@ -205,7 +206,7 @@ class TwilioConnection extends StateMachine {
           // Do nothing.
           break;
         case 'heartbeat':
-          this._handleHeartbeat();
+          // Do nothing.
```
@makarandp0 You can remove this case altogether I think.

makarandp0 replied:
I think keeping this case with `// do nothing` comment helps with understaning possible `message.type`  values.

manjeshbhargav commented on lib/twilioconnection.js:
```diff
@@ -303,11 +305,18 @@ class TwilioConnection extends StateMachine {
-    const heartbeatTimeout = negotiatedTimeout + HEARTBEAT_TIMEOUT_OFFSET;
+
+    const { maxConsecutiveMissedHeartbeats } = this._options;
+    // we will essentially wait for (maxConsecutiveMissedHeartbeats) missed heartbeats before closing the connection.
+    const heartbeatTimeout = (negotiatedTimeout * maxConsecutiveMissedHeartbeats);
```
```suggestion

    // NOTE(mpatwardhan): Wait for 3 missed heartbeats before closing the TwilioConnection.
    const heartbeatTimeout = negotiatedTimeout * maxConsecutiveMissedHeartbeats;
```

manjeshbhargav commented on lib/twilioconnection.js:
```diff
@@ -303,11 +305,18 @@ class TwilioConnection extends StateMachine {
+    // we will essentially wait for (maxConsecutiveMissedHeartbeats) missed heartbeats before closing the connection.
+    const heartbeatTimeout = (negotiatedTimeout * maxConsecutiveMissedHeartbeats);
+
+    // Outgoing heartbeat should be sent 200 ms earlier than the specified interval to account for network latency.
+    const outgoingHeartbeatTimeout = negotiatedTimeout - OUTGOING_HEARTBEAT_OFFSET;
```
```suggestion
    // NOTE(mpatwardhan): Outgoing heartbeat is sent 200 ms earlier to account for network latency.
    const outgoingHeartbeatTimeout = negotiatedTimeout - OUTGOING_HEARTBEAT_OFFSET;
```

ceaglest commented on lib/twilioconnection.js:
```diff
@@ -435,7 +445,7 @@ class TwilioConnection extends StateMachine {
  * @typedef {object} TwilioConnectionOptions
  * @property {LogLevel} [logLevel=warn] - Log level of the {@link TwilioConnection}
- * @property {number} [maxConsecutiveMissedHeartbeats=5] - Max. number of consecutive "heartbeat" messages that can be missed
+ * @property {number} [maxConsecutiveMissedHeartbeats=3] - Max. number of (effective) consecutive "heartbeat" messages that can be missed
```
Is the 5 just a documentation a typo, while DEFAULT_MAX_CONSECUTIVE_MISSED_HEARTBEATS represents the correct value of 3?

makarandp0 replied:
yes, seems like just documentation typo.

ceaglest commented on lib/twilioconnection.js:
```diff
@@ -30,14 +30,16 @@ const states = {
 const DEFAULT_MAX_CONSECUTIVE_MISSED_HEARTBEATS = 3;
 const DEFAULT_MAX_REQUESTED_HEARTBEAT_TIMEOUT = 5000;
+const DEFAULT_OPEN_TIMEOUT = 15000;
 const DEFAULT_WELCOME_TIMEOUT = 5000;
```
Welcomes should be allowed for up to 15 seconds, and the hello should be sent up to 3 times at [0, 5, 10] seconds. It seems like this change should be part of this PR rather than https://github.com/twilio/twilio-video.js/pull/985 (JSDK-2777).

makarandp0 replied:
Looks like #985 already covered that.

makarandp0 replied:
looks like #985 already handled this requirement.

ceaglest commented on lib/twilioconnection.js:
```diff
@@ -197,6 +195,9 @@ class TwilioConnection extends StateMachine {
       }
+
+      // Each incoming message should be treated as an incoming heartbeat
+      this._handleHeartbeat();
```
From our discussion on slack I don't think you should treat each sent or received message as a heartbeat. I was planning on rolling back this change in C++ since it is out of sync with the Server and we don't have agreement between teams. The concern I have is that treating every message as a heartbeat allows the connection to become further and further behind.

makarandp0 replied:
@ceaglest I understand that we shouldn't treat outgoing message as heartbeat - because server expects heartbeats. But don't see why not for incoming messages. If we miss heartbeat but received incoming message why should we disconnect ?
Can you elaborate more on 
>  treating every message as a heartbeat allows the connection to become further and further behind.

makarandp0 replied:
resolved after the offline discussion - we decided to treat every incoming message as incoming heartbeat, and outgoing message as outgoing heartbeat. since server already [does the same](https://code.hq.twilio.com/twilio/twilio-connections/blob/master/connections-core/src/main/java/com/twilio/connections/actor/ServerConnectionActor.java#L559) -

makarandp0 replied:
resolved after offline discussion - we decided to treat incoming message as heartbeat.

ceaglest commented:
FWIW, it looks like treating every message as a heartbeat was attempted in the Java implementation but the [PR](https://code.hq.twilio.com/twilio/twilio-connections/pull/63) was not merged.

makarandp0 commented:
@ceaglest and @manjeshbhargav - Thank you for the discussion on this earlier.  I have re-introduce "treat outgoing message as heartbeat" change. Please take another look. I am planning to verify that server indeed is okay with us treating outgoing message as heartbeat before merging this.

ceaglest commented on lib/twilioconnection.js:
```diff
@@ -267,6 +277,9 @@ class TwilioConnection extends StateMachine {
       }
+
+      // Each incoming message should be treated as an incoming heartbeat
+      this._handleHeartbeat();
```
Calling `_handleHeartbeat` for every `message.type` might result in unexpected logs from the SDK. Move this call into `heartbeat` and `msg` instead?

makarandp0 replied:
Good point - I have updated now to call this only for msg and heartbeat cases.

makarandp0 commented:
hey @ceaglest, I have addressed your feedback, can you please take another look.

makarandp0 commented:
I verified that
-  if client does not send any heartbeats but send messages (track enable/disable) at regular interval. server does not disconnect.
- if client does not send any heartbeats and not messages as well. server does disconnect the connection after few missed heartbeats.

This verifies our assumption that treating messages as heartbeat is okay with server. I will merge this now.

makarandp0 commented:
update the signaling websocket connection heartbeat logic to:

- Each outgoing message is now treated as an outgoing heartbeat.
- Each incoming message is treated as an incoming heartbeat.
- The incoming heartbeat timeout is now 15 seconds, .
- Removed the offset added to the incoming heartbeat timeout.
- added offset of -200ms  200 to outgoing timeout to account for network latency.
- Add a timeout for getting the onopen event on the WebSocket (15 sec). After that, attempt to reconnect only if this is not the first attempt.
